src/Service/Bulk/BulkCreateUsersService.php

```diff
@@ -0,0 +1,294 @@
+        BulkUserCreationCsvWriter $csvWriter,
+        Filesystem $filesystem,
+        string $generatedUsersFilePath
+    ) {
+        $this->lineItemRepository = $lineItemRepository;
+        $this->assignmentRepository = $assignmentRepository;
+        $this->ltiInstanceLoadBalancerFactory = $ltiInstanceLoadBalancerFactory;
+        $this->csvWriter = $csvWriter;
+        $this->generatedUsersFilePath = $generatedUsersFilePath;
+        $this->filesystem = $filesystem;
+    }
+
+    /**
+     * @throws LineItemNotFoundException
+     */
+    public function createUsers(
+        array $lineItemIds,
+        array $lineItemSlugs,
+        array $userPrefixes,
+        int $batchSize,
+        ?string $groupPrefix
+    ): UserCreationResult {
+
+        $notExistLineItemsArray = $userGroupIds = [];
+
+        if (!empty($lineItemIds) || !empty($lineItemSlugs)) {
+            $this->lineItemIds = $lineItemIds;
+            $this->lineItemSlugs = $lineItemSlugs;
+            $criteria = $this->getFindLineItemCriteria();
+            $lineItems = $this->lineItemRepository->findLineItemsByCriteria($criteria)->jsonSerialize();
+            if (empty($lineItems)) {
+                $exceptionMessage = $this->lineItemIds
+                    ? implode(',', $this->lineItemIds) . ' Line item id(s) not exist in the system'
+                    : implode(',', $this->lineItemSlugs) . ' Line item slug(s) not exist in the system';
+                throw new LineItemNotFoundException($exceptionMessage);
+            }
+
+            $this->generateSlugData($lineItems);
+            $notExistLineItemsArray = $this->lineItemIds
+                ? array_diff($lineItemIds, array_keys($this->lineItemSlugs))
+                : array_diff($lineItemSlugs, array_values($this->lineItemSlugs));
+        }
+
+        if (empty($lineItemIds) && empty($lineItemSlugs)) {
+            $this->getAllLineItemSlugs();
+        }
+
+        $userNameLastIndexes = $this->getLastUserAssignedToLineItems($this->lineItemSlugs);
+
+        $userGroupAssignCount = 0;
+        if ($groupPrefix) {
+            $loadBalancer = $this->ltiInstanceLoadBalancerFactory->__invoke('userGroupId');
+            $userGroupIds = $loadBalancer->generateGroupIds($groupPrefix);
+            $userGroupAssignCount = (int) ceil(
+                count($userPrefixes) * count($this->lineItemSlugs) * $batchSize / count($userGroupIds)
+            );
+        }
+
+        $slugTotalUsers = $this->createBulkUserAssignmentData(
+            $userNameLastIndexes,
+            $userGroupIds,
+            $userPrefixes,
+            $userGroupAssignCount,
+            $groupPrefix,
+            $batchSize
+        );
+
+        $message = $this->getOperationResultMessage($slugTotalUsers, $userPrefixes);
+
+        return new UserCreationResult($message, $notExistLineItemsArray);
+    }
+
+    public function getOperationResultMessage(array $slugTotalUsers, array $userPrefix): string
```

Comment:
this method has nothing to do with the service

Reply:
As suggested by @luisfn getOperationResultMessage() method is used to create the response message in a report format to show the total number of users created on each slug for user prefix

Reply:
it violates SRP 
please move it to dedicated service, maybe a normalizer